### PR TITLE
Implement performUpgrade flag

### PR DESCRIPTION
**Pull request by kickash32**
This PR implements #1992. It introduces the performUpgrade, which behaves similar to the forceUpgrade flag but only runs the world upgrade if the world was not upgraded in the current MC version. The information is stored inside papermeta.yml file inside each world's folder.

### CODE REVIEW

File: Spigot-Server-Patches/0400-Implement-performUpgrade-flag.patch [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+     private boolean forceUpgrade;
++    private boolean performUpgrade; // Paper
+     private boolean eraseCache;
+     private float at;
+     public final Executor executorService;
+@@ -167,6 +168,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
+     public final double[] recentTps = new double[ 3 ];
+     public final SlackActivityAccountant slackActivityAccountant = new SlackActivityAccountant();
+     // Spigot end
++    String worldMetaFile = "papermeta.yml"; // Paper
```
Comment (kashike) [inline Spigot-Server-Patches/0400-Implement-performUpgrade-flag.patch]: `paper-state.yml` or something

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by kickash32** [comment]
The data is now stored in level.dat of each world.

**Review by MiniDigger — COMMENTED** [review]
I think Paper prefixes all nbt values with Paper to avoid potential future collisions so Paper.LastForceUpgrade should be used.

**Comment by kickash32** [comment]
Done.

**Comment by aikar** [comment]
this seems way more complicated than it needs to be. the worlds level.dat should already have a data version stored in it, which you can use to determine if it needs to be converted or not w/o any new paper value.

**Comment by kickash32** [comment]
That NBT tag, from my understanding, only stores the latest version that loaded the world. It has little to do with the fact if all the chunks were upgraded or not. I do agree this should be simplified, unfortunately, this reading/writing of the level.dat needs to be done before the world is fully loaded so I cannot use the methods that already exist. I suppose it should be possible to delay writing of the level.dat to a later time which may simplify the code.

**Comment by aikar** [comment]
Oh right yeah it could be updated w/o entire world being force upgraded. so I guess it is needed for you silly people who insist on using force upgrade lol.

**Comment by Tsoccerguy3** [comment]
We have run into this issue with our hosting server at Fragnet , but it's not just them its most hosting services . They do not provide any way to update game versions . This has been a issue since 1.3 was released . Just today we submitted a ticket to Fragnut and they agreed to add the switch manually , but have to wait in line , as all the minecraft server owners were requesting the same thing . I guess now the hosts will build this feature into there WEB GUI . why didnt anyone come up with this idea when 1.3 came out .? Been suffering from bad performance for more than a year , because chunks were updating on the fly .

Great Idea though !

**Comment by kickash32** [comment]
updated for 1.14.4

**Comment by Tsoccerguy3** [comment]
https://github.com/PaperMC/Paper/issues/1992#issuecomment-515094853

**Comment by Tsoccerguy3** [comment]
Im requesting that this feature be merged, Downloading 4 gig server to do the java paper.jar --forceUpgrade on a local computer and then uploading the converted map is a big waste of time and money . A lot of people use a shared host and are not allowed access to add switches to the server boot

**Comment by electronicboy** [comment]
This patch also uses a flag...

**Comment by Intelli** [comment]
Reason this is useful (even though it’s still a flag) is stated here: https://github.com/PaperMC/Paper/issues/1992#issuecomment-486418617

**Comment by BillyGalbreath** [comment]
I personally don't think this is needed at all. It's not like you have to perform an upgrade to your world on startup. The `--forceUpgrade` is already added as a convenience.

**Comment by cyberkitsune** [comment]
If this is deemed as necessary, why not have it as a config option in paper.yml? That circumvents any possibility of server hosts not allowing flags. Additionally, that brings it in line with most of paper’s patches that support toggling.

**Comment by Intelli** [comment]
—forceUpgrade may be considered a convenience feature, but if you’ve tried playing on a large map/server without running the upgrade first, you’ll find that performance is absolutely abysmal.

That is, I’d find it hard to recommend that  server owners upgrade from a < 1.13 map without running the flag.

**Comment by Tsoccerguy3** [comment]
A toggle in a config file ,sounds like a great idea , this would satisfy not having access to the Hosts low level boot commands.

**Comment by BillyGalbreath** [comment]
> —forceUpgrade may be considered a convenience feature, but if you’ve tried playing on a large map/server without running the upgrade first, you’ll find that performance is absolutely abysmal.
> 
> That is, I’d find it hard to recommend that server owners upgrade from a < 1.13 map without running the flag.

Maybe I'm just having a hard time understanding what this PR is supposed to solve. There is no way to know per-chunk if a chunk has already been converted or not. The`--forceUpgrade` simply loads every single chunk at server startup. This PR adds a new flag `--performUpgrade` that simply loads every single chunk on server startup.. Where is the benefit?

Checking the level.dat if the world has been upgraded or not already won't do anything new..

**Comment by Intelli** [comment]
> This PR adds a new flag --performUpgrade that simply loads every single chunk on server startup.

That isn't what this does... 

https://github.com/PaperMC/Paper/pull/2149

**Comment by BillyGalbreath** [comment]
> > This PR adds a new flag --performUpgrade that simply loads every single chunk on server startup.
> 
> That isn't what this does...
> 
> #2149

Maybe I'm just blind, but that [is indeed what this PR does](https://github.com/PaperMC/Paper/pull/2149/files#diff-fabef32f569d85989ad4a225ec725095R35).. If the server admin has the access to add the `--performUpgrade` flag, then they have the access to use the `--forceUpgrade` flag and then remove it when finished. This PR solves nothing from what I can see.

**Comment by electronicboy** [comment]
The diference between --force is that it checks if the process has already been ran on the world by storing the version in the level.dat file vs always running

**Comment by BillyGalbreath** [comment]
> The diference between --force is that it checks if the process has already been ran on the world by storing the version in the level.dat file vs always running

I understand that.. But why? It's just as easy to set the flag when needed and remove it when not needed. This just seems like a lazy thing imho. And it won't even help those that can't set their own flags anyway..

**Comment by Tsoccerguy3** [comment]
This solves the issue if the server host , example Fragnet does not allow anyone access to the java parameters and the switches . having this PR merged allows for a toggle in the config file such as --forceUpgrade =true or false . some server operators with their own dedicated hardware would just do java -jar paper.jar --forceUpgrade  one time and the chunks would be converted to the next version . Not every server owner hosts on there own hardware .

**Comment by electronicboy** [comment]
This PR does not add anything to the config files...

**Comment by Tsoccerguy3** [comment]
I know that it does not add to the config, It was proposed up higher to revise the PR to make it a optional toggle in the paper.yml

**Comment by BillyGalbreath** [comment]
> This solves the issue if the server host , example Fragnet does not allow anyone access to the java parameters and the switches . having this PR merged allows for a toggle in the config file...

Uh, wrong. This PR adds another flag that you won't be able to use.

Don't get me wrong, I'm all for helping the little guy, but this isn't it. If it was a config option then I'd be all for it and have no objections. But as it is, this doesn't help anyone with anything except being lazy.

**Comment by Tsoccerguy3** [comment]
> 
> 
> This PR implements #1992. It introduces the performUpgrade, which behaves similar to the forceUpgrade flag but only runs the world upgrade if the world was not upgraded in the current MC version. The information is stored inside papermeta.yml file inside each world's folder.

Does this PR convert a maps chunks from 1.14.4 to 1.15 without asking a server hosting company to manually run , java -jar paper.jar --forceUpgrade  1 time to force all maps and all chunks are safely updated to the next version ?

**Comment by electronicboy** [comment]
This PR still requires you to add a flag to your startup args in order to upgrade the worlds

**Comment by Tsoccerguy3** [comment]
Then it needs to be revised so that the switch is a toggle in the paper.yml . Not everyone has access to add the switch to startup . I was already told by Fragnet that all Minecraft servers with them share the same startup

**Comment by Intelli** [comment]
Having it as an option in paper.yml sounds good to me also.

**Pull request closed without merging**